File: 프로그래머스/2/12952. N－Queen/N－Queen.py
def solution(n):
    # 조합을 저장할 set
    visit = set()
    def check(i,j,col,d1,d2):
        # 행, 열, 대각선 집합에 n이 속하는지
        if j in col: return False
        if (i-j) in d1: return False
        if (i+j) in d2: return False
        return True
    
    # 행, 1-n**2 list, 금지열, 금지대각선 2가지
    def dfs(i,path,col,d1,d2):
        # 만약 idx == n and len(path) == n이면 visit에 추가
        if i == n and len(path) == n:
            key = frozenset(path)
            visit.add(key)
            return
        # idx행에서 j열에 q을 둘 수 있는지 판단
        for j in range(n):
            if check(i,j,col,d1,d2):
                col.add(j)
                d1.add((i-j))
                d2.add((i+j))
                dfs(i+1,path+[(i,j)],col,d1,d2)
                col.remove(j)
                d1.remove((i-j))
                d2.remove((i+j))
        # idx행에 둘 수 있는 곳이 없다면 return
        return
    dfs(0,[],set(),set(),set())
    return len(visit)

File: 프로그래머스/2/12952. N－Queen/N－Queen.kt
class Solution {
    private var m = 0
    private var ans = 0
    fun dfs(i: Int, col: MutableSet<Int>, d1: MutableSet<Int>, d2: MutableSet<Int>) {
        if (i == m) {
            ans++
            return
        }
        
        for (j in 0 until m) {
            if (j !in col && (i-j) !in d1 && (i+j) !in d2) {
                col.add(j)
                d1.add(i-j)
                d2.add(i+j)
                dfs(i+1,col,d1,d2)
                col.remove(j)
                d1.remove(i-j)
                d2.remove(i+j)
            }
        }
        return
    }
    
    fun solution(n: Int): Int {
        m = n
        dfs(0,mutableSetOf(),mutableSetOf(),mutableSetOf())
        return ans
    }
}